packages/@lwc/engine-core/src/framework/modules/attr-unless-prop.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import {
+    htmlAttributeToProperty,
+    isNull,
+    isUndefined,
+    StringCharCodeAt,
+    XML_NAMESPACE,
+    XLINK_NAMESPACE,
+} from '@lwc/shared';
+
+import { unlockAttribute, lockAttribute } from '../attributes';
+import { RendererAPI } from '../renderer';
+import { EmptyObject } from '../utils';
+import { VBaseElement } from '../vnodes';
+
+const ColonCharCode = 58;
+
+export function patchAttrUnlessProp(
+    oldVnode: VBaseElement | null,
+    vnode: VBaseElement,
+    renderer: RendererAPI
+) {
+    const {
+        data: { attrs },
+        elm,
+    } = vnode;
+    if (isUndefined(elm) || isUndefined(attrs)) {
+        return;
+    }
+
+    const { removeAttribute, setAttribute } = renderer;
+    const oldAttrs = isNull(oldVnode) ? EmptyObject : oldVnode.data.attrs;
+
+    for (const name in attrs) {
+        const cur = attrs[name];
+        const old = oldAttrs[name];
+
+        const propName = htmlAttributeToProperty(name);
+        if (propName in elm) {
+            (elm as any)[propName] = cur;
+        } else if (old !== cur) {
```

Comment:
we should do nothing if the old is same as cur, so, why is the block above still carry on?

Reply:
If the custom element was upgraded since the last render and the property now exists, the property should be set.

Reply:
Offline conversation: we don't make assumptions about whether the component is designed to take attributes or props, as long as we set it on the component one way or another. if they want the value on upgrade, they need to support attributes.